### PR TITLE
Improvements to ReadOnlyBytes

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
cc: @shiftylogic, @ahsonkhan, @davidfowl 

Added more through tests for slicing
Found a bug and fixed it
Optimized slicing to empty ROBs

### CODE REVIEW

File: src/System.Buffers.Experimental/System/Buffers/ReadOnlyBytes.cs [hunk 1]
```diff
@@ -129,9 +135,20 @@ public ReadOnlyBytes Slice(int index)
             else {
                 var first = First;
                 if (first.Length > index) {
-                    return new ReadOnlyBytes(first.Slice(index), _rest);
+                    var slice = first.Slice(index);
+                    if(slice.Length > 0 || _rest!=null) {
+                        return new ReadOnlyBytes(slice, _rest);
```
Comment (ahsonkhan) [inline src/System.Buffers.Experimental/System/Buffers/ReadOnlyBytes.cs]: _rest can be null?

Comment (KrzysztofCwalina) [inline src/System.Buffers.Experimental/System/Buffers/ReadOnlyBytes.cs]: Yes. If the multi-segment structure contains just one segment.

Comment (ahsonkhan) [inline src/System.Buffers.Experimental/System/Buffers/ReadOnlyBytes.cs]: I see. To confirm, this should be an 'or' (as in we could end up with ReadOnlyBytes (0, _rest) or ReadOnlyBytes(slice, null)? Both are valid?
```C#
if(slice.Length > 0 || _rest!=null) { 
```

Comment (KrzysztofCwalina) [inline src/System.Buffers.Experimental/System/Buffers/ReadOnlyBytes.cs]: Yes, it should be an OR. If slice.length is zero but _rest is not null, we don't want to return an empty ROB. Same if slice is longer than zero.

**Comment by davidfowl** [comment]
The formatting is all messed up 😄

**Pull request merged**